pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/TriggerSomaticResultSurveyRoute.java

```diff
@@ -0,0 +1,149 @@
+                    requestPayload.getParticipantId(),
+                    triggerId,
+                    triggerUpdatedSomaticUpload.getBucket(),
+                    triggerUpdatedSomaticUpload.getBlobPath());
+            return DDPRequestUtil.triggerFollowupSurvey(instance, payloadToSend, requestPayload.getSurveyName());
+        } else {
+            throw new DSMBadRequestException("Bad somatic document id.  Not triggering followup");
+        }
+    }
+
+    private boolean isEligibleToBeSent(SomaticResultUpload resultUpload) {
+        if (Boolean.FALSE.equals(resultUpload.getIsVirusFree())) {
+            throw new DSMBadRequestException(
+                    "The file selected is either not finished being scanned for viruses or was removed because one was found.");
+        }
+        if (resultUpload.getDeletedAt() > 0) {
+            throw new DSMBadRequestException("The file selected was already deleted and cannot be sent.");
+        }
+        return true;
+    }
+
+    private long buildTriggerComment(String userIdRequest) {
+        long currentTime = System.currentTimeMillis();
+        try {
+            return addTriggerCommentIntoDB(userIdRequest, currentTime);
+        } catch (Exception ex) {
+            throw new DsmInternalError("Error encountered adding trigger and comment, please contact a DSM developer.");
+        }
+    }
+
+    //this is modeled after the Query parameters used in TriggerSurveyRoute.java because we want this endpoint
+    //to be similar to that one, but add the support for the required documentId parameter that we need to look
+    //up the correct somatic file that was uploaded.
+    private SomaticResultTriggerRequestPayload getRequestPayload(QueryParamsMap queryParams, Request request) {
+        SomaticResultTriggerRequestPayload result = new SomaticResultTriggerRequestPayload();
+        if (queryParams.value("surveyName") != null) {
+            result.setSurveyName(queryParams.get("surveyName").value());
+        } else {
+            throw new DSMBadRequestException("No surveyName query param was sent");
+        }
+        if (!result.getSurveyName().equalsIgnoreCase("SOMATIC_RESULTS")) {
+            throw new DSMBadRequestException("This route only works with SOMATIC_RESULTS surveys.");
+        }
+        if (queryParams.value("surveyType") != null) {
+            result.setSurveyType(queryParams.get("surveyType").value());
+        } else {
+            throw new DSMBadRequestException("No surveyType query param was sent");
+        }
+        if (!result.getSurveyType().equalsIgnoreCase("REPEATING")) {
+            throw new DSMBadRequestException("This route only supports REPEATING types");
+        }
+        try {
+            JsonObject jsonObject = new JsonParser().parse(request.body()).getAsJsonObject();
+            String ddpParticipantId = jsonObject.getAsJsonObject().get("participantId").getAsString();
```

Comment:
Do we need to check for null or validate that this is a valid ID? If not, it might be worth a comment about why.

Reply:
The ID, the participant and the realm are all used as the key to the entry in the database query invoked by line 50 (service.getSomaticResultByIdPtptAndRealm).  If the row doesn't have all three matching, the update will fail because a row will not be found.